python/ray/dag/compiled_dag_node.py

```diff
@@ -1253,32 +1199,56 @@ def _get_or_compile(
-                reader_handles_set = set()
+                # A dictionary that maps an InputNode or InputAttributeNode to its
+                # readers and the node on which the reader is running. Use `set` to
+                # deduplicate readers on the same actor because with CachedChannel
+                # each actor will only read from the shared memory once.
+                input_dag_node_to_reader_and_node_set: Dict[
+                    Union[InputNode, InputAttributeNode],
+                    Set[Tuple["ray.actor.ActorHandle", str]],
+                ] = defaultdict(set)
+
                 for idx in task.downstream_task_idxs:
                     reader_task = self.idx_to_task[idx]
                     assert isinstance(reader_task.dag_node, ClassMethodNode)
                     reader_handle = reader_task.dag_node._get_actor_handle()
-                    if reader_handle not in reader_handles_set:
-                        reader_and_node_list.append(
-                            (reader_handle, self._get_node_id(reader_handle))
-                        )
-                    reader_handles_set.add(reader_handle)
-                task.output_channels = [
-                    do_allocate_channel(
+                    reader_node_id = self._get_node_id(reader_handle)
+                    for arg in reader_task.args:
+                        if isinstance(arg, InputAttributeNode) or isinstance(
+                            arg, InputNode
+                        ):
+                            input_dag_node_to_reader_and_node_set[arg].add(
+                                (reader_handle, reader_node_id)
+                            )
+
+                # A single channel is responsible for sending the same data to
+                # corresponding consumers. Therefore, we create a channel for
+                # each InputAttributeNode, or a single channel for the entire
+                # input data if there are no InputAttributeNodes.
+                task.output_channels = []
+                for input_dag_node in input_dag_node_to_reader_and_node_set:
+                    reader_and_node_list = list(
+                        input_dag_node_to_reader_and_node_set[input_dag_node]
+                    )
+                    output_channel = do_allocate_channel(
                         self,
                         reader_and_node_list,
                         typ=type_hint,
                     )
-                ]
+                    task.output_channels.append(output_channel)
+                    task.output_idxs.append(
+                        None
+                        if isinstance(input_dag_node, InputNode)
+                        else input_dag_node.key
+                    )
+
+                    # Update the InputAttributeNode's `output_channels`, which is
+                    # used to determine whether to create a CachedChannel.
+                    if isinstance(input_dag_node, InputAttributeNode):
```

Comment:
can you give me more details about what this is doing? I am a little confused because the task has output channel, but then we have separate input attr task, and they have its own output channels? why is that?

Reply:
1. This ensures consistent behavior with `ClassMethodNode` when `is_class_method_output` is set to True.

https://github.com/ray-project/ray/blob/35b453445b49ad286c08099953c000c3942acf6a/python/ray/dag/compiled_dag_node.py#L1241-L1253

2. When creating a `CachedChannel`, we need to handle `InputAttributeNode` separately because its output channel is empty, which will cause the assertion `assert len(upstream_task.output_channels) == 1` to fail.

https://github.com/ray-project/ray/blob/35b453445b49ad286c08099953c000c3942acf6a/python/ray/dag/compiled_dag_node.py#L1343-L1356